Use `BorrowedMappedPages` to replace the usage of `owning_ref` to stably-deref types overlaid atop `MappedPages`

kevinaboos opened the pull request:
We no longer need `owning_ref` (which is unsound, see #482) for this purpose with the `MappedPages` type. This also avoid needing to wrap a deref'd `MappedPages` object in a `Box`, which is a wasteful heap allocation pattern.

kevinaboos committed:
Added 3 new types that represent re-typed regions of memory based on an underlying MappedPages object, for easier usage of MappedPages.

kevinaboos committed:
Merge branch 'theseus_main' into embedded_mapped_pages

kevinaboos committed:
Remove old `EmbeddedMappedPages` which is infeasible.

Simplify implementations of `Borrowed[Mut]MappedPages` based on
`MappedPages::as_type[_mut]()` functions.

kevinaboos committed:
Add docs

kevinaboos committed:
avoid unnecessary usage of `NonNull::dangling()`

Co-authored-by: Klim Tsoutsman <klimusha@gmail.com>

kevinaboos committed:
Add forgotten `mut` to fix build errors

kevinaboos committed:
Merge branch 'theseus_main' into embedded_mapped_pages

kevinaboos committed:
Alignment is now properly checked by `MappedPages::as_type[_mut]()`

kevinaboos commented:
Note: this is not yet complete and is blocked on #648

kevinaboos committed:
* Improve `BorrowedMappedPages`
* Add `BorrowedSliceMappedPages`
* Actually used `BorrowedMappedPages` in a few crates:
  * framebuffer
  * apic
  * hpet

kevinaboos commented:
Reminder to properly utilize the sealed trait for the marker structs that implement the `Mutability` trait

kevinaboos committed:
Use `BorrowedMappedPages` in the e1000 and ixgbe NIC drivers

kevinaboos committed:
cleanup, plus iommu and ioapic

kevinaboos committed:
apply to mlx NIC driver

kevinaboos committed:
Remove unneeded dependencies on `owning_ref`

kevinaboos committed:
remove `owning_ref` from `ethernet_smoltcp_device` crate

kevinaboos committed:
Fix the sealed trait implementation for `Mutability`

kevinaboos committed:
Merge branch 'theseus_main' into use_borrowed_mapped_pages

kevinaboos committed:
fix imports in memory and paging

kevinaboos commented:
@tsoutsman requesting your review since we discussed the initial PR #648. I've extended and modified things quite a bit since that first PR. 

@Ramla-I requesting your review since this makes several changes to NIC drivers.

tsoutsman commented on kernel/memory/src/paging/mapper.rs:
```diff
@@ -919,100 +920,221 @@ pub fn mapper_from_current() -> Mapper {
+        length: usize,
+    ) -> Result<BorrowedSliceMappedPages<T>, (MappedPages, &'static str)> {
+        let borrowed_mp = Self {
+            ptr: match mp.as_slice::<T>(byte_offset, length) {
```
From the [`slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) docs:
> The total size len * mem::size_of::<T>() of the slice must be no larger than isize::MAX. See the safety documentation of [pointer::offset](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset).

`as_slice` uses `usize::checked_mul` to compute `size_in_bytes`. We should ensure that it also fits in an `isize`.

I can't imagine a case where it wouldn't but we may as well if we're checking for overflows.

kevinaboos replied:
thanks. Resolved by #674

tsoutsman commented on kernel/memory/src/paging/mapper.rs:
```diff
@@ -919,100 +920,221 @@ pub fn mapper_from_current() -> Mapper {
+        length: usize,
+    ) -> Result<Self, (MappedPages, &'static str)> {
+        let borrowed_mp = Self {
+            ptr: match mp.as_slice_mut::<T>(byte_offset, length) {
```
Same as above but for `as_slice_mut`.

kevinaboos replied:
thanks. Resolved by #674

tsoutsman reviewed (changes requested):
Aside from some small problems, it looks good.

Can we shorten the function names to just `new` or something rather? The structs already have long names and I think the long initialiser names are unnecessary.

The structs should implement `AsRef`, `Borrow`, and `BorrowMut` (obviously only the mutable variants). However, for the `Borrow` and `BorrowMut` implementations to be correct, we'll have to forward the `Eq`, `Hash`, and `Ord` implementations to the `T` ( or `[T]`). `Rc`, `Box`, etc. do this.

As an alternative API:
```rust
struct BorrowedMappedPages<Ref>
where
    Ref: Reference,
{
    ptr: Unique<Ref::Target>,
    mp: MappedPages,
    _marker: PhantomData<Ref>,
}

impl<T> BorrowedMappedPages<&T>
where
    T: zerocopy::FromBytes,
{
    pub fn new() -> Self {
        todo!();
    }
}

impl<T> BorrowedMappedPages<&mut T>
where
    T: zerocopy::FromBytes,
{
    pub fn new() -> Self {
        todo!();
    }
}

impl<T> BorrowedMappedPages<&[T]>
where
    T: zerocopy::FromBytes,
{
    pub fn new() -> Self {
        todo!();
    }
}

impl<T> BorrowedMappedPages<&mut [T]>
where
    T: zerocopy::FromBytes,
{
    pub fn new() -> Self {
        todo!();
    }
}
pub trait Reference: private::Sealed {
    type Target;
}

impl<T> Reference for &T
where
    T: zerocopy::FromBytes,
{
    type Target = T;
}
impl<T> Reference for &mut T
where
    T: zerocopy::FromBytes,
{
    type Target = T;
}
impl<T> Reference for &[T]
where
    T: zerocopy::FromBytes,
{
    type Target = T;
}
impl<T> Reference for &mut [T]
where
    T: zerocopy::FromBytes,
{
    type Target = T;
}

impl<T> private::Sealed for &T where T: zerocopy::FromBytes {}
impl<T> private::Sealed for &mut T where T: zerocopy::FromBytes {}
impl<T> private::Sealed for &[T] where T: zerocopy::FromBytes {}
impl<T> private::Sealed for &mut [T] where T: zerocopy::FromBytes {}

mod private {
    pub trait Sealed {}
}
```
I don't like it because you can't generalise over mutability and it's confusing to have the same struct represent both references and slices, but I figured I might as well throw it out there for _inspiration_ 💡. ✨. 💭. 🤩. 💫.

kevinaboos committed:
Merge branch 'theseus_main' into use_borrowed_mapped_pages

kevinaboos committed:
Shorten constructors, add convenience methods for `MappedPages`

kevinaboos committed:
Implement key traits for `BorrowedMappedPages` and `BorrowedSliceMappedPages`:
* `AsRef`, `AsMut`, `Borrow`, `BorrowMut`
* `PartialEq`, `Eq`
* `PartialOrd`, `Ord`
* `Hash`
* `Deref` and `DerefMut` were already implemented

kevinaboos commented:
> Can we shorten the function names to just `new` or something rather? The structs already have long names and I think the long initialiser names are unnecessary.

Yep, good idea. I changed them to `from()` and `from_mut()`, and I added some convenience methods that operate on `MappedPages`:
* `into_borrowed()`
* `into_borrowed_mut()`
* `into_borrowed_slice()`
* `into_borrowed_slice_mut()`

Those function names are a bit long but it's unavoidable since we need to disambiguate between the 8 related variants, the four `as*` functions and the four `into_borrrowed*` functions.

> The structs should implement `AsRef`, `Borrow`, and `BorrowMut` (obviously only the mutable variants). However, for the `Borrow` and `BorrowMut` implementations to be correct, we'll have to forward the `Eq`, `Hash`, and `Ord` implementations to the `T` ( or `[T]`). `Rc`, `Box`, etc. do this.

done.

kevinaboos committed:
cleanup

kevinaboos committed:
Merge branch 'theseus_main' into use_borrowed_mapped_pages
